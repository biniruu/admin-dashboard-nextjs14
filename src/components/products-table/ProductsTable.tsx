'use client'

import Image from 'next/image'
import Link from 'next/link'

import styles from './productsTable.module.css'

import type { Product } from 'types'
import { deleteProduct } from 'utils/actions'

const tableHead: string[] = ['title', 'description', 'price', 'created at', 'stock', 'action']

interface Props {
  products: Product[]
}

function ProductsTable({ products }: Props) {
  return (
    <table className={`${styles.table} w-full`}>
      <thead>
        <tr>
          {tableHead.map(item => (
            <td key={item}>{item}</td>
          ))}
        </tr>
      </thead>
      <tbody>
        {products?.map(product => {
          const { id, img, title, desc, price, createdAt, stock } = product

          return (
            <tr key={id}>
              <td>
                <div className="flex items-center gap-default">
                  <Image
                    src={img || '/noproduct.jpg'}
                    alt=""
                    width={40}
                    height={40}
                    className="rounded-half object-cover"
                  />
                  {title}
                </div>
              </td>
              <td>{desc}</td>
              <td>${price}</td>
              <td>{createdAt?.toString().slice(4, 16)}</td>
              <td>{stock}</td>
              <td>
                <div className="flex gap-default">
                  <Link href={`/dashboard/products/${id}`}>
                    <button className={`${styles.button} bg-[teal]`}>View</button>
                  </Link>
                  <form action={deleteProduct}>
                    <input type="hidden" name="id" value={id} />
                    <button className={`${styles.button} bg-[crimson]`}>Delete</button>
                  </form>
                </div>
              </td>
            </tr>
          )
        })}
      </tbody>
    </table>
  )
}

export default ProductsTable
