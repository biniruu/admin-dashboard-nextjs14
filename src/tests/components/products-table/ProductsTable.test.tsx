import renderer from 'react-test-renderer'

import ProductsTable from '@components/products-table/ProductsTable'
import type { Product } from 'types'

jest.mock('lib/actions', () => ({
  deleteProduct: jest.fn(),
}))

describe('ProductsTable component', () => {
  let products: Product[]

  beforeEach(() => {
    products = [
      {
        id: '1',
        img: '/product1.jpg',
        title: 'Product 1',
        desc: 'Description 1',
        price: 10,
        createdAt: new Date(),
        updatedAt: new Date(),
        stock: 5,
      },
      {
        id: '2',
        img: '/product2.jpg',
        title: 'Product 2',
        desc: 'Description 2',
        price: 20,
        createdAt: new Date(),
        updatedAt: new Date(),
        stock: 10,
      },
    ]
  })

  test('should render the component', () => {
    const result = renderer.create(<ProductsTable products={products} />).toJSON()

    expect(result).toMatchSnapshot()
  })
})
