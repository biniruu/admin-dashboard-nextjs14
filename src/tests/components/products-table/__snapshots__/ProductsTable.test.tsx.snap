// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsTable component should render the component 1`] = `
<table
  className="table w-full"
>
  <thead>
    <tr>
      <td>
        title
      </td>
      <td>
        description
      </td>
      <td>
        price
      </td>
      <td>
        created at
      </td>
      <td>
        stock
      </td>
      <td>
        action
      </td>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <div
          className="flex items-center gap-default"
        >
          <img
            alt=""
            className="rounded-half object-cover"
            data-nimg="1"
            decoding="async"
            height={40}
            loading="lazy"
            onError={[Function]}
            onLoad={[Function]}
            src="/_next/image?url=%2Fproduct1.jpg&w=96&q=75"
            srcSet="/_next/image?url=%2Fproduct1.jpg&w=48&q=75 1x, /_next/image?url=%2Fproduct1.jpg&w=96&q=75 2x"
            style={
              {
                "color": "transparent",
              }
            }
            width={40}
          />
          Product 1
        </div>
      </td>
      <td>
        Description 1
      </td>
      <td>
        $
        10
      </td>
      <td>
        Apr 14 2024 
      </td>
      <td>
        5
      </td>
      <td>
        <div
          className="flex gap-default"
        >
          <a
            href="/dashboard/products/1"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onTouchStart={[Function]}
          >
            <button
              className="button bg-[teal]"
            >
              View
            </button>
          </a>
          <form
            action={[MockFunction]}
          >
            <input
              name="id"
              type="hidden"
              value="1"
            />
            <button
              className="button bg-[crimson]"
            >
              Delete
            </button>
          </form>
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div
          className="flex items-center gap-default"
        >
          <img
            alt=""
            className="rounded-half object-cover"
            data-nimg="1"
            decoding="async"
            height={40}
            loading="lazy"
            onError={[Function]}
            onLoad={[Function]}
            src="/_next/image?url=%2Fproduct2.jpg&w=96&q=75"
            srcSet="/_next/image?url=%2Fproduct2.jpg&w=48&q=75 1x, /_next/image?url=%2Fproduct2.jpg&w=96&q=75 2x"
            style={
              {
                "color": "transparent",
              }
            }
            width={40}
          />
          Product 2
        </div>
      </td>
      <td>
        Description 2
      </td>
      <td>
        $
        20
      </td>
      <td>
        Apr 14 2024 
      </td>
      <td>
        10
      </td>
      <td>
        <div
          className="flex gap-default"
        >
          <a
            href="/dashboard/products/2"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onTouchStart={[Function]}
          >
            <button
              className="button bg-[teal]"
            >
              View
            </button>
          </a>
          <form
            action={[MockFunction]}
          >
            <input
              name="id"
              type="hidden"
              value="2"
            />
            <button
              className="button bg-[crimson]"
            >
              Delete
            </button>
          </form>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
